File: packages/core/src/auth/guard.ts
import type { Authorizer } from "./authorizer.js";
import { AuthorizeUnavailableError, type HeaderLike, type Principal } from "./types.js";

export type GuardOutcome =
  | { status: 200; principal: Principal | null }
  | { status: 401 | 403 | 503; principal: null; message: string };

/**
 * The identify-then-authorize sequence every binding runs, with the outcome
 * already mapped onto the status code that binding returns.
 *
 * An absent identity is authorized as anonymous, not refused here: the server's
 * `anyone` short-circuit is what admits a public interface.
 */
export async function guard(
  authz: Authorizer,
  headers: HeaderLike,
): Promise<GuardOutcome> {
  const principal = await authz.identify(headers);

  try {
    const decision = await authz.authorize(principal);
    if (!decision.allowed) {
      return principal
        ? { status: 403, principal: null, message: "Forbidden" }
        : { status: 401, principal: null, message: "Unauthorized" };
    }
    return { status: 200, principal };
  } catch (err) {
    if (err instanceof AuthorizeUnavailableError) {
      return {
        status: 503,
        principal: null,
        message: "Authorization unavailable",
      };
    }
    throw err;
  }
}

File: packages/core-py/src/astropods_adapter_core/auth/guard.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Mapping, Optional

from .authorizer import Authorizer
from .types import AuthorizeUnavailableError, Principal


@dataclass
class GuardOutcome:
    status: int
    principal: Optional[Principal] = None
    message: str = ""


def guard(authz: Authorizer, headers: Mapping[str, Any]) -> GuardOutcome:
    """The identify-then-authorize sequence every binding runs, with the outcome
    already mapped onto the status code that binding returns.

    An absent identity is authorized as anonymous, not refused here: the
    server's ``anyone`` short-circuit is what admits a public interface.
    """
    principal = authz.identify(headers)

    try:
        decision = authz.authorize(principal)
    except AuthorizeUnavailableError:
        return GuardOutcome(status=503, message="Authorization unavailable")

    if not decision.allowed:
        if principal is None:
            return GuardOutcome(status=401, message="Unauthorized")
        return GuardOutcome(status=403, message="Forbidden")
    return GuardOutcome(status=200, principal=principal)


async def guard_async(authz: Authorizer, headers: Mapping[str, Any]) -> GuardOutcome:
    """``guard`` off the event loop. Only a cache miss reaches the network, so
    most calls never leave the thread pool's fast path."""
    return await asyncio.to_thread(guard, authz, headers)
